src/cudamatrix/cu-matrix.cc

```diff
@@ -1700,6 +1700,96 @@ double TraceMatMat(const CuMatrixBase<double> &A,
 
+template<typename Real>
+void AddMatMatBatched(const Real alpha, std::vector<CuSubMatrix<Real>* > &C,
+		const std::vector<CuSubMatrix<Real>* > &A, MatrixTransposeType transA,
+		const std::vector<CuSubMatrix<Real>* > &B, MatrixTransposeType transB,
+		const Real beta) {
+    // the batch counts should be the same	
+    int32 batchCountA = static_cast<int32>(A.size()), batchCountB = static_cast<int32>(B.size()), batchCountC = static_cast<int32>(C.size());
+    KALDI_ASSERT((batchCountA == batchCountB) && (batchCountB == batchCountC));
+    int32 batchCount = batchCountA;
+
+    if (batchCount == 0) return;
+    
+    // all elements must have the same num-rows, num-cols and stride
+    for (int32 i = 0; i < batchCount-1; i++) {
+      KALDI_ASSERT(A[i]->NumRows() == A[i+1]->NumRows());
+      KALDI_ASSERT(A[i]->NumCols() == A[i+1]->NumCols());
+      KALDI_ASSERT(A[i]->Stride() == A[i+1]->Stride());
+      KALDI_ASSERT(B[i]->NumRows() == B[i+1]->NumRows());
+      KALDI_ASSERT(B[i]->NumCols() == B[i+1]->NumCols());
+      KALDI_ASSERT(B[i]->Stride() == B[i+1]->Stride());
+      KALDI_ASSERT(C[i]->NumRows() == C[i+1]->NumRows());
+      KALDI_ASSERT(C[i]->NumCols() == C[i+1]->NumCols());
+      KALDI_ASSERT(C[i]->Stride() == C[i+1]->Stride());
+    }
+    // CUBLAS is col-major, cudamatrix is row-major, how to do the mapping?
+    // keep trans..., just swap A&B matrices: A->B B->A
+    MatrixIndexT m = ((transB==kTrans)? B[0]->NumRows() : B[0]->NumCols()); 
+    MatrixIndexT n = ((transA==kTrans)? A[0]->NumCols() : A[0]->NumRows());
+    MatrixIndexT k = ((transB==kTrans)? B[0]->NumCols() : B[0]->NumRows());
+    MatrixIndexT k1 = ((transA==kTrans)? A[0]->NumRows() : A[0]->NumCols());
+
+    KALDI_ASSERT(m == C[0]->NumCols());
+    KALDI_ASSERT(n == C[0]->NumRows());
+    KALDI_ASSERT(k == k1);
+
+    if (m == 0) return;
+
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    // place all data (A[batchCount], B[batchCount], C[batchCount]) in memory together
+    // in order to make only one call of cudaMemcpy 	  
+    Real **device_abc_array = (Real**)CuDevice::Instantiate().Malloc(3*batchCount*sizeof(Real*));
+    const Real **device_a_array = const_cast<const Real**>(device_abc_array); 
+    const Real **device_b_array = const_cast<const Real**>(device_abc_array) + batchCount;
+    Real **device_c_array = device_abc_array + 2 * batchCount;
+    const Real **host_abc_array = new const Real*[3*batchCount];
+    const Real **host_a_array = host_abc_array; 
+    const Real **host_b_array = host_abc_array + batchCount; 
+    const Real **host_c_array = host_abc_array + 2 * batchCount; 
+    
+    for (int32 i = 0; i < batchCount; i++) {
+      host_a_array[i] = A[i]->data_;
+      host_b_array[i] = B[i]->data_;
+      host_c_array[i] = C[i]->data_;
+    }
+
+    CU_SAFE_CALL(cudaMemcpy(device_abc_array, host_abc_array, 3*batchCount*sizeof(Real*), cudaMemcpyHostToDevice));
+
+    Timer tim;
```

Comment:
BTW, there doesn't need to be any fancy logarithmic reduction  in
AddMatBlocks-- just compute the sum using a simple for-loop.
Dan


On Wed, Sep 9, 2015 at 4:48 PM, Daniel Povey <dpovey@gmail.com> wrote:

>
>>
>>    1.
>>
>>    at line 3902 of nnet2/nnet-component.cc in function Propagate(),
>>    should the last argument of AddVecToRows() be 1.0 instead of 0.0?
>>
>> It doesn't matter, since that is the scale on something that's zero
> anyway.
>
>>
>>    1.
>>
>>    at line 4178 in function Update(), filters_grad is used to accumulate
>>    the sum. What I am doing right now is to compute each term of sum in
>>    parallel using AddMatMatBatched(), and then sum over the resulting matrices
>>    together to compute the final filters_grad. I was wondering if it is what I
>>    was supposed to do.
>>
>> Firstly, make sure you allocate a single large matrix, with a large
> number of rows (possibly better for alignment reasons), rather than
> separately allocating a bunch of small matrices.
> Rather than adding the small matrices together manually using repeated
> calls to AddMatMat, please write a CuMatrix function as follows:
>
>   /// This function is like AddMat (does *this += alpha * src), but
> handles the case
>   /// where src.NumRows() may be a multiple of this->NumRows(); in that
> case
>   /// it divides 'src' into blocks of size equal to '*this' and adds each
> of them (times alpha)
>   /// to *this.
>   void AddMatBlocks(Real alpha, const CuMatrixBase<Real> &src);
>
> This can call a custom kernel- you can modify it from the one that AddMat
> currently calls.  For now it's not handling transposing, or the case where
> the num-cols is also a multiple, so it shouldn't be too hard to write.
>
> You can then use this function to avoid the loop over AddMat.
> Add some test code for AddMatBlocks in cu-matrix-test.cc, and speed test
> for it in cu-matrix-speed-test.cc.
>
> For the bias_grad, you don't have to do anything with a loop, you can just
> compute it as follows using a single call:
>  bias_grad.AddRowSumMat(1.0, out_deriv, 1.0);
>
> Testing will automatically happen in the nnet-component-test.cc, you
> shouldn't have to write anything extra.  But make your changes to the
> forward and backward propagation one by one rather than at the same time,
> and make sure the tests run-- because the test compare the forward and
> backward passes and check the derivatives, it's easier to change things one
> at a time so you know where the problem is.
>
> Dan
>
>
>
>
>>
>>    1.
>>
>>    Could you please tell me how to test my code?
>>
>> Thanks,
>> Yiming
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/127/files#r39091954>.
>>
>
>